ddtrace/appsec/iast/overhead_control_engine.py

```diff
@@ -0,0 +1,82 @@
+"""
+The Overhead control engine (OCE) is an element that by design ensures that the overhead does not go over a maximum
+limit. It will measure operations being executed in a request and it will deactivate detection
+(and therefore reduce the overhead to nearly 0) if a certain threshold is reached.
+"""
+import os
+import threading
+from typing import TYPE_CHECKING
+
+
+if TYPE_CHECKING:  # pragma: no cover
+    from typing import Set
+    from typing import Type
+
+MAX_REQUESTS = int(os.environ.get("DD_IAST_MAX_CONCURRENT_REQUESTS", 2))
+MAX_VULNERABILITIES_PER_REQUEST = int(os.environ.get("DD_IAST_VULNERABILITIES_PER_REQUEST", 2))
+
+
+class Operation(object):
+    """Common operation related to Overhead Control Engine (OCE). Every vulnerabilities/taint_sinks should inherit
+    from this class. OCE instance requests for these methods to control the overhead in each request.
+    """
+
+    _lock = threading.Lock()
+    _quota = MAX_VULNERABILITIES_PER_REQUEST
+
+    @classmethod
+    def reset(cls):
+        cls._quota = MAX_VULNERABILITIES_PER_REQUEST
+
+    @classmethod
+    def decrement_quota(cls):
+        if cls._quota > 0:
+            cls._quota -= 1
```

Comment:
I believe this should also be controlled with a lock.

How about a single function `acquire_quota` that checks for `has_quota` and `decrements_quota` if there is quota available?

Reply:
Also: I insist in renaming this `quota` to `vulnerability_quota` and `request` to `request_quota`